### PR TITLE
Added Bounds Tests

### PR DESCRIPTION
**Pull request by Chris3606**
# Changes
- Minor refactor in `KnownSeriesRandomTests` which allows us to avoid duplicating an entire function to support decimal function testing
- Created facilities in the unit tests project for creating one of each type of generator in the library.  Consolidating this to a single function allows the actual unit test classes to share the code, which will limit the amount of modification needed when new generator types are added
- Added a framework to test all `IEnhancedRandom` implementations to ensure they respect the bounds defined by the contract for the interface's functions.
    - The theories involved create a fairly large number of test cases, so it will take a moment to run

# Goals/End-State
The primary goal here is to create an extensible framework that tests the generator implementations within the library, to ensure that they implement the `IEnhancedRandom` contract properly.  I elected to use `KnownSeriesRandom` in the testing process, in order to ensure the bounds checking code already written for it was effectively shared so it wasn't duplicated.

Hopefully, this should allow future generators and modifications to take place with confidence that the `IEnhancedRandom` contract is properly maintained, and as well that the `KnownSeriesRandom` implementation will stay in sync.

### CODE REVIEW

File: ShaiRandom.UnitTests/DataGenerators.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        /// return a <see cref="KnownSeriesRandom"/>, since that generator is a special case.
+        /// </summary>
+        /// <remarks>
+        /// New generator implementations should be added here to ensure they're unit tested.
+        /// </remarks>
+        /// <param name="includeWrappers">Whether or not to include things in the ShaiRandom.Wrappers namespace.</param>
+        /// <returns>All qualifying generator types in the library.</returns>
+        public static IEnumerable<IEnhancedRandom> CreateGenerators(bool includeWrappers)
+        {
+            yield return new DistinctRandom();
```
Comment (tommyettinger) [inline ShaiRandom.UnitTests/DataGenerators.cs:23]: These use random seeds. This shouldn't be a problem in the vast majority of situations, but there is always a small chance of any distribution-dependent tests failing out of the blue because the seed happened to be in a range that would not have the expected distribution. For example, we don't have the Mersenne Twister here, but it's mathematically guaranteed to have a seed that produces the result `4` 623 times in a row as its first outputs. We need to just be aware that some PRs that depend on all tests passing may fail randomly with a low but non-zero likelihood.

A way around this would be to retry failing distribution-dependent tests with a fixed seed, like `1`, which should only be a problem if the 1-seed also fails. It might also be useful to serialize the generators this returns and print or log the appropriate serialized generator on a failure, so it can be reproduced. That's all something that can wait until much later.

File: ShaiRandom.UnitTests/KnownSeriesRandomTests.cs [hunk 2]
```diff
@@ -272,7 +273,11 @@ private void TestFloatingFunctionExclusiveBounds<T>(string nameOfFunctionToTest,
             dynamic zero = (T)Convert.ChangeType(0.0, typeof(T));
             T pointOne = (T)Convert.ChangeType(0.1, typeof(T));
             T pointTwo = (T)Convert.ChangeType(0.2, typeof(T));
-            T epsilon = (T)typeof(T).GetField("Epsilon")!.GetValue(null)!;
+            T epsilon = pointOne switch
+            {
+                decimal _ => (T)Convert.ChangeType(0.0000000000000000000000000001M, typeof(T)),
+                _ => (T)typeof(T).GetField("Epsilon")!.GetValue(null)!
+            };
```
Comment (tommyettinger) [inline ShaiRandom.UnitTests/KnownSeriesRandomTests.cs:280]: This is much better than what I had, heh.

**Review by tommyettinger — APPROVED** [review]
Looks good. I'm not sure what to do about the random seeds, because they could cause PRs to fail mysteriously if a particular seed was used that happened to break expectations. Using fixed seeds might not adequately reveal issues in some places, though... And some fixed seeds will necessarily produce badly-distributed output for some section of their output. I'll merge this as-is and see how often the tests fail as a whole, I guess.
